### DISCUSSION TITLE
Mooncake performance

### DISCUSSION THREAD

**Discussion by SamuelBrand1** [TuringLang/Turing.jl#2418] [General]
I saw there was a general thread for Zygote performance, so it seemed logical to have one for mooncake.

Running locally (Mac with M1 cores, julia 1.11.1) the linear regression example https://turinglang.org/docs/tutorials/05-linear-regression/ takes about x15 the time compared to either `AutoForwardDiff` or `AutoReverseDiff(compile=true)`.

I'm happy to do a MWE on this but its simply the tutorial code with extra `ReverseDiff` and `Mooncake` deps and a different `adtype` kwarg passed to `NUTS` constructor.

Any immediate ideas about where the performance hit it happening here?

**Comment by torfjelde** [comment]
@willtebbutt !

I also encountered this, and I believe we concluded it was likely due to a lack of custom rules for Mooncake.jl (these are currently somewhat annoying and life, for the most part, in DistributionsAD.jl). So technically, nothing should be stopping us from achieving the same perf here. Would be useful to confirm this though by implementing the corresponding AD rules.

**Reply by SamuelBrand1** [reply]
Thats interesting @torfjelde . Is there a way to get `build_rrule` to work here in a way that is compatible with the model declaration?

**Reply by torfjelde** [reply]
IIUC it's more about just writing a rrule for logpdf of MvNormal that works with MoonCakejl; no need to worry about the model itself:)

**Reply by SamuelBrand1** [reply]
Yeah, but I don't want to get doomed to writing `rrule`s every time I've had a new idea...

**Reply by torfjelde** [reply]
I think there's some confusion here.

There should be nothing stopping Mooncake.jl from being _at least_ as performant as ReverseDiff.jl in addition to being a) more flexible (since it does not rely on `TrackedArray`, which can, unless you're careful, accidentally be converted into a `Array{TrackedReal}`, which tanks performance in many cases), and b) supporting mutation, which in turn allows further perf optimisations.

But we've spent years working around issues with ReverseDiff.jl, etc. and some of that included writing custom rrules for specific logpdf computations; in particular, we've done this _when handcoded rrules are better_. Mooncake.jl however, has seen no such treatment. And so there are cases where we have optimised ReverseDiff.jl to a point where it's not really a fair comparison to Mooncake.jl _at the moment_.

Hence, once we start finding these cases where a custom rrule is useful, e.g. the `MvNormal` case, then we'll add similar rules for Mooncake.jl. But this should not be the norm; custom rrules is expected to be the exception rather than the rule.

Buuuut @willtebbutt should really comment more generally on this; he might also know what's up with this particular example and whether just copying the rrule we have in DistributionsAD.jl would be sufficient to match it or if something else is missing.

Btw, just to clarify @SamuelBrand1 , when you say "15 slower", are you talking about benchmarking the model with the AD backends or are you talking about the `sample` call itself?

**Reply by willtebbutt** [reply]
Ah, right, so I think this is actually [this](https://github.com/TuringLang/DynamicPPL.jl/pull/727) issue -- in short Turing.jl isn't currently caching rules properly, which will cause really quite poor performance in this case. So, sadly, I think I accidentally told an untruth in the chat the other day @SamuelBrand1 -- Mooncake.jl is _almost_ ready for use with Turing.jl. models via `AutoMooncake`.

We're currently waiting on the general registry working properly again in order to get this merged, but having checked out to it locally, I see the following results:
```julia
julia> benchmark_model(model, adbackends=[AutoForwardDiff(), AutoReverseDiff(true), AutoMooncake(; config=nothing)])
2-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "evaluation" => 2-element BenchmarkTools.BenchmarkGroup:
          tags: []
          "linked" => Trial(541.000 ns)
          "standard" => Trial(417.000 ns)
  "gradient" => 3-element BenchmarkTools.BenchmarkGroup:
          tags: []
          "AutoReverseDiff(compile=true)" => 2-element BenchmarkTools.BenchmarkGroup:
                  tags: ["ReverseDiff [compiled]"]
                  "linked" => Trial(4.666 μs)
                  "standard" => Trial(4.000 μs)
          "AutoForwardDiff()" => 2-element BenchmarkTools.BenchmarkGroup:
                  tags: ["ForwardDiff"]
                  "linked" => Trial(1.541 μs)
                  "standard" => Trial(1.334 μs)
          "AutoMooncake{Nothing}(nothing)" => 2-element BenchmarkTools.BenchmarkGroup:
                  tags: ["AutoMooncake{Nothing}(nothing)"]
                  "linked" => Trial(4.833 μs)
                  "standard" => Trial(4.375 μs)
```
This is using the `model` variable constructed in https://turinglang.org/docs/tutorials/05-linear-regression/ .
If you don't have this PR, you'll see quite a lot more overhead because the rule is getting re-built each time.

To be honest, I'm still a little bit disappointed that we're slower than ReverseDiff.jl here (I'm fine with not beating ForwardDiff.jl for such a small problem), but I would have expected to be somewhere between ReverseDiff.jl and ForwardDiff.jl. I've done some local profiling and a surprisingly large amount of time is being spent in the reverse-pass associated to `LinearAlgebra.BLAS.gemv!` -- I've not spent a lot of time optimising this particular hand-coded rule, so I should probably just improved its performance, and hopefully Mooncake.jl will be beat ReverseDiff.jl.

Note that what Tor says about things from DistributionsAD.jl. is correct (Mooncake.jl doesn't use it at all), however, it is also true that it doesn' have to. In short, I have no intention at all of incorporating rules from DistributionsAD.jl.

**Reply by willtebbutt** [reply]
With both the DynamicPPL PR and [this](https://github.com/compintell/Mooncake.jl/pull/407) Mooncake PR to fix gemv, you should see something like the following:
```julia
julia> benchmark_model(
           model,
           adbackends=[
               AutoForwardDiff(),
               AutoReverseDiff(compile=true),
               Turing.ADTypes.AutoMooncake(; config=nothing)
           ],
       )
2-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "evaluation" => 2-element BenchmarkTools.BenchmarkGroup:
          tags: []
          "linked" => Trial(583.000 ns)
          "standard" => Trial(416.000 ns)
  "gradient" => 3-element BenchmarkTools.BenchmarkGroup:
          tags: []
          "AutoReverseDiff(compile=true)" => 2-element BenchmarkTools.BenchmarkGroup:
                  tags: ["ReverseDiff [compiled]"]
                  "linked" => Trial(4.666 μs)
                  "standard" => Trial(4.000 μs)
          "AutoForwardDiff()" => 2-element BenchmarkTools.BenchmarkGroup:
                  tags: ["ForwardDiff"]
                  "linked" => Trial(1.500 μs)
                  "standard" => Trial(1.333 μs)
          "AutoMooncake{Nothing}(nothing)" => 2-element BenchmarkTools.BenchmarkGroup:
                  tags: ["AutoMooncake{Nothing}(nothing)"]
                  "linked" => Trial(3.542 μs)
                  "standard" => Trial(3.125 μs)
```
These PRs ought really to get merged tomorrow 🤞 .

**Reply by torfjelde** [reply]
Amazin' @willtebbutt :eyes:

**Reply by SamuelBrand1** [reply]
> Btw, just to clarify @SamuelBrand1 , when you say "15 slower", are you talking about benchmarking the model with the AD backends or are you talking about the sample call itself?

The whole sample call (on second run to avoid compilation time).

**Reply by SamuelBrand1** [reply]
> Ah, right, so I think this is actually https://github.com/TuringLang/DynamicPPL.jl/pull/727 issue -- in short Turing.jl isn't currently caching rules properly, which will cause really quite poor performance in this case. So, sadly, I think I accidentally told an untruth in the chat the other day @SamuelBrand1 -- Mooncake.jl is almost ready for use with Turing.jl. models via AutoMooncake.

I thought this was happening but didn't have time to burrow in.

**Reply by SamuelBrand1** [reply]
> Note that what Tor says about things from DistributionsAD.jl. is correct (Mooncake.jl doesn't use it at all), however, it is also true that it doesn' have to. In short, I have no intention at all of incorporating rules from DistributionsAD.jl.

Yes, this was the impression I was getting and super happy its going this way!

**Reply by seabbs** [reply]
Any interest in us throwing a few simple models together to use in benchmarking or some co-creation?

**Reply by willtebbutt** [reply]
Very interested. Did you have anything in particular in mind?
